x/gamm/twap/store_test.go

```diff
@@ -0,0 +1,191 @@
+func (s *TestSuite) TestTrackChangedPool() {
+	tests := map[string][]uint64{
+		"single":         {1},
+		"duplicated":     {1, 1},
+		"four":           {1, 2, 3, 4},
+		"many with dups": {1, 2, 3, 4, 3, 2, 1},
+	}
+	for name, test := range tests {
+		s.Run(name, func() {
+			// Test that no cumulative ID registers as tracked
+			changedPools := s.twapkeeper.GetChangedPools(s.Ctx)
+			s.Require().Empty(changedPools)
+
+			// Track every pool in list
+			cumulativeIds := map[uint64]bool{}
+			for _, v := range test {
+				cumulativeIds[v] = true
+				s.twapkeeper.TrackChangedPool(s.Ctx, v)
+			}
+
+			changedPools = s.twapkeeper.GetChangedPools(s.Ctx)
+			s.Require().Equal(len(cumulativeIds), len(changedPools))
+			for _, v := range changedPools {
+				s.Require().True(cumulativeIds[v])
+			}
+			s.Commit()
+		})
+	}
+}
+
+// TestGetAllMostRecentRecordsForPool takes a list of records as test cases,
+// and runs storeNewRecord for everything in sequence.
+// Then it runs GetAllMostRecentRecordsForPool, and sees if its equal to expected
+func (s *TestSuite) TestGetAllMostRecentRecordsForPool() {
+	tPlusOne := baseTime.Add(time.Second)
+	baseRecord := newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA")
+	tPlusOneRecord := newEmptyPriceRecord(1, tPlusOne, "tokenB", "tokenA")
+	tests := map[string]struct {
+		recordsToSet    []types.TwapRecord
+		poolId          uint64
+		expectedRecords []types.TwapRecord
+	}{
+		"set single record": {
+			recordsToSet:    []types.TwapRecord{baseRecord},
+			poolId:          1,
+			expectedRecords: []types.TwapRecord{baseRecord},
+		},
+		"query non-existent pool": {
+			recordsToSet:    []types.TwapRecord{baseRecord},
+			poolId:          2,
+			expectedRecords: []types.TwapRecord{},
+		},
+		"set single record, different pool ID": {
+			recordsToSet:    []types.TwapRecord{newEmptyPriceRecord(2, baseTime, "tokenB", "tokenA")},
+			poolId:          2,
+			expectedRecords: []types.TwapRecord{newEmptyPriceRecord(2, baseTime, "tokenB", "tokenA")},
+		},
+		"set two records": {
+			recordsToSet:    []types.TwapRecord{baseRecord, tPlusOneRecord},
+			poolId:          1,
+			expectedRecords: []types.TwapRecord{tPlusOneRecord},
+		},
+		"set two records, reverse order": {
+			// The last record, independent of time, takes precedence for most recent.
+			recordsToSet:    []types.TwapRecord{tPlusOneRecord, baseRecord},
+			poolId:          1,
+			expectedRecords: []types.TwapRecord{baseRecord},
+		},
+		"set multi-asset pool record": {
+			recordsToSet: []types.TwapRecord{
+				newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenB"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenA")},
+			poolId: 1,
+			expectedRecords: []types.TwapRecord{
+				newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenA"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenB")},
```

Comment:
Was the order of records here switched on purpose or does it just not matter? Just asking to ensure I fully understand.

Reply:
The returned order is the order that our helper for getting all pairs returns